index.js

```diff
@@ -0,0 +1,414 @@
+/*!
+ * Copyright (c) 2018-2019 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+const axios = require('axios');
+const bedrock = require('bedrock');
+const BigNumber = require('bignumber.js');
+
+const logger = require('./logger.js');
+const paymentService = require('bedrock-payment');
+
+// list of all PayPal Supported currency codes.
+const currencies = require('./currencies.js');
+
+const {PaymentStatus, Errors} = paymentService;
+
+const {config} = bedrock;
+const {BedrockError} = bedrock.util;
+
+// payment refers to a Payment.
+// You can find Payment in bedrock-web-payment:
+// @see https://github.com/digitalbazaar/bedrock-web-payment/
+
+const getConfig = () => {
+  const {api, clientId, secret} = config.paypal;
+  if(!clientId) {
+    throw new BedrockError(
+      'Missing PayPal clientId', Errors.Data);
+  }
+
+  if(!secret) {
+    throw new BedrockError(
+      'Missing PayPal secret', Errors.Data);
+  }
+
+  if(!api) {
+    throw new BedrockError(
+      'Missing PayPal api', Errors.Data);
+  }
+
+  return {api, clientId, secret};
+};
+
+// credentials needed on the front end to make payments.
+const getGatewayCredentials = () => {
+  const {clientId} = getConfig();
+  if(!clientId) {
+    throw new BedrockError(
+      'No Credentials Found for PayPal',
+      Errors.NotFound, {public: true, httpStatusCode: 404}
+    );
+  }
+  return {service: 'paypal', paypal_client_id: clientId};
+};
+
+/**
+ * Format Amount - Takes in an amount and formats it.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.amount - An amount.
+ *
+ * @throws DataError - If the currency is not supported by PayPal.
+ * @throws DataError - If the amount's value is not a number.
+ *
+ * @returns {object} A formatted amount.
+ */
+const formatAmount = ({amount}) => {
+  const supported = currencies.supported.has(amount.currency_code);
+  if(!supported) {
+    throw new BedrockError(
+      `Unsupported PayPal currency ${amount.currency_code}`,
+      Errors.Data, {public: true}
+    );
+  }
+  const bigAmount = BigNumber(amount.value);
+  const notNumber = bigAmount.toString() === 'NaN';
+  if(notNumber) {
+    throw new BedrockError(
+      `Invalid amount ${amount.value} ${amount.currency_code}`,
+      Errors.Data, {public: true}
+    );
+  }
+  if(currencies.noDecimal.has(amount.currency_code)) {
+    amount.value = bigAmount.toFixed(0).toString();
+    return amount;
+  }
+  amount.value = bigAmount.toFixed(2).toString();
+  return amount;
+};
+
+/**
+ * Gets a PayPal auth token so we can call on the api.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} [options.clientId = paypal_client_id] - PayPal Client Id.
+ * @param {string} [options.secret = paypal_secret] - PayPal Secret.
+ *
+ * @returns {Promise<object>} The data returned with an `access_token`.
+ */
+const getAuthToken = async ({clientId, secret, api}) => {
```

Comment:
Looks like many calls to other APIs lead to a new request for a token.  Do these tokens have a TTL or something, are there any opportunities for cacheing the token instead of adding this round-trip to every request?

Reply:
Good point. I actually think caching the token here would be better.

Reply:
ok good idea. Implementation could maybe use some work. Let me know: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/9f162230b10c3c0a9d15a79784ef35c417eea2a4

Reply:
@mattcollier Did the https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/9f162230b10c3c0a9d15a79784ef35c417eea2a4 commit address your concerns?